### ISSUE TITLE
fix newlines on linux/mac

### ISSUE THREAD

**Issue by konsumer** [MikeMnD/tiled-to-godot-export#6]
This is from discussion on #5 

I am on Linux, and it seems to be interpreting newlines in map file differently:

![Screenshot from 2020-04-14 13-34-28](https://user-images.githubusercontent.com/83857/79272646-988ebd00-7e56-11ea-8d7d-ad324caa5ddf.png)

**Comment by konsumer** [comment]
I tried out this tester, as-per @MikeMnD suggestion, and it works:
[tiled-godot-master-working.zip](https://github.com/MikeMnD/tiled-to-godot-export/files/4477707/tiled-godot-master-working.zip)

![Screenshot from 2020-04-14 13-49-36](https://user-images.githubusercontent.com/83857/79272962-089d4300-7e57-11ea-91b0-6ce2447b7790.png)

**Comment by konsumer** [comment]
I am happy to test on OSX (Mojave) too.

**Comment by MikeMnD** [comment]
Ok, but still we have to make some change to the generation of the new line in the export script.
So it should work on all OS.
Could you try adding locally " \"   before the blank lines in the scripts.
And reexport and check if the blank lines are generated fine.
Something like this:
![image](https://user-images.githubusercontent.com/186191/79273759-3660a700-7eac-11ea-8bad-f25434688cfe.png)

It's info i found here: https://stackoverflow.com/questions/38000659/template-strings-es6-prevent-line-breaks

If this doesn't fix it I'll load up Ubuntu tomorrow and will try to fix it there

**Comment by konsumer** [comment]
we could also replace `\n\r` with `\n` on output. might make it nicer to read....

**Comment by MikeMnD** [comment]
I'm ok if it doesn't make problems on windows :P

**Comment by konsumer** [comment]
Hmm, I tried something like this in both exporters, and it still fails:

``` writeToFile() {
        // noinspection JSUnresolvedVariable
        const file = new TextFile(this.fileName, TextFile.WriteOnly);
        // log(this.tileset.imageWidth, this.tileset.imageHeight);
        let tilesetTemplate = this.getTilesetTemplate();
        file.write(tilesetTemplate.replace(/\s{2,}/g, ""));
        file.commit();
    }```

I'm wondering if even the fact that I have git-cloned it is stripping the special windows characters. Maybe we need to do it more old-school: `lines.join('\n\r')` where lines is an array.

**Comment by MikeMnD** [comment]
Did you try adding a backslash in the template literal - like I showed in the screenshot above  line 115 for example?

**Comment by konsumer** [comment]
> Did you try adding a backslash in the template literal - like I showed in the screenshot above line 115 for example?

Nah, I tired to make it more readable and not break anyone's code-style by just doing a replace on file-write. That way we get all the same code, except 1 line changes. I will play with my regex. So, I actually want linefeeds in godot files?

> I just liked the template literals cause they preserve the formatting... but yeah

Yeh, I am a huge fan of ES6-templates.

**Comment by konsumer** [comment]
I still can't figure it out. I did this:

```js
.replace(/\r\n|\n/g, "\r\n")
```

And it still doesn't work.

**Comment by bjorn** [comment]
What is the problem here? Does Godot need a specific style of newlines to be able to read its files correctly?

The `TextFile` object opens the file in text mode, which generally makes any `\n` translate to system-specific line endings. So, a newline will write as `\n` on Linux and macOS but as `\r\n` on Windows.

I think if you need to avoid this automatic conversion then you need to use `BinaryFile`, though I'm currently unsure if the API enables writing strings to it. Question is how to handle the text codec in this case.

**Comment by konsumer** [comment]
> What is the problem here? Does Godot need a specific style of newlines to be able to read its files correctly?

It seemed that way, but now I'm not sure. Once I started using the dev build of tiled it seemed to go away. Maybe it's related to the imageHeight/imageWidth bug. I can't seem to reproduce on the dev build.

Newlines was one difference between the 2 files, but I think the important difference was that the non-working output had different tile-numbers.

**Comment by konsumer** [comment]
I think this issue can also just be closed. I think the answer to both my issues is simply "it only works in the development version of tiled".

**Comment by MikeMnD** [comment]
Well, i was a bit busy today and felt sorry that I couldn't test it on Linux.
Glad to hear that the problem was to an older version of Tiled.
I've added to the readme that the plugins require 1.3.4 as minimum version.

**Comment by bjorn** [comment]
@konsumer This problem was resolved in Tiled 1.3.4 or the development version? I would have no idea what change or fix affected that. :S

**Comment by konsumer** [comment]
Yep. My guess is that it's not newlines, it's that 1.33 does not have imagHeight/imageWidth, and there are some loops that depend on that to get the tile IDs and column-counts right. In my own files, the main diff between running the plug-in on 1.33 and 1.34 seems to be the outputted tile ids.
